opm/autodiff/ParallelDebugOutput.hpp

```diff
@@ -0,0 +1,553 @@
+                for( auto it = localWellState_.wellMap().begin(); it != end; ++it )
+                {
+                    const std::string& name = it->first;
+                    const int wellIdx = it->second[ 0 ];
+
+                    // write well name
+                    writeString( buffer, name );
+
+                    // write well data
+                    buffer.write( localWellState_.bhp()[ wellIdx ] );
+                    buffer.write( localWellState_.thp()[ wellIdx ] );
+                    const int wellRateIdx = wellIdx * localWellState_.numPhases();
+                    for( int np=0; np<localWellState_.numPhases(); ++np )
+                        buffer.write( localWellState_.wellRates()[ wellRateIdx + np ] );
+
+                    // TODO: perfRates and perfPress, need to figure out the index
+                    // mapping there.
+                }
+            }
+
+            void readWells( MessageBufferType& buffer )
+            {
+                int nWells = -1;
+                buffer.read( nWells );
+                // unpack all wells that have been sent
+                std::string name ;
+                for( int well = 0; well < nWells ; ++well )
+                {
+                    // read well name for local identification
+                    readString( buffer, name );
+
+                    // unpack values
+                    auto it = globalWellState_.wellMap().find( name );
+                    if( it == globalWellState_.wellMap().end() )
+                    {
+                        OPM_THROW(std::logic_error,"global state does not contain well " <<  name );
+                    }
+                    const int wellIdx = it->second[ 0 ];
+
+                    buffer.read( globalWellState_.bhp()[ wellIdx ] );
+                    buffer.read( globalWellState_.thp()[ wellIdx ] );
+                    const int wellRateIdx = wellIdx * globalWellState_.numPhases();
+                    for( int np=0; np<globalWellState_.numPhases(); ++np )
+                        buffer.read( globalWellState_.wellRates()[ wellRateIdx + np ] );
+
+                    // TODO: perfRates and perfPress, need to figure out the index
+                    // mapping there.
+                }
+            }
+
+        };
+
+        // gather solution to rank 0 for EclipseWriter
+        bool collectToIORank( const SimulatorState& localReservoirState,
+                              const WellState& localWellState )
+        {
+            PackUnPackSimulatorState packUnpack( localReservoirState, globalReservoirState_,
+                                                 localWellState, globalWellState_,
+                                                 localIndexMap_, indexMaps_, isIORank() );
+
+            //toIORankComm_.exchangeCached( packUnpack );
+            toIORankComm_.exchange( packUnpack );
+#ifndef NDEBUG
+            // mkae sure every process is on the same page
+            toIORankComm_.barrier();
```

Comment:
Because the communication is essentially non-blocking, e.g. all ranks send and then proceed with computations while only the I/O rank waits until all data is received and then writes. That how it has to be for this to be efficient (besides the bottleneck that all send to one). Therefore, the barrier would take away this benefit and is only there for debugging purpose, i.e. not defined NDEBUG ;-).